qmi/core/transport.py

```diff
@@ -723,18 +738,126 @@ def discard_read(self) -> None:
         self._safe_socket.settimeout(0)
         while True:
             try:
-                b = self._safe_socket.recv(4096)
-            except socket.timeout:
+                b, addr = self._safe_socket.recvfrom(4096)
+            except (BlockingIOError, socket.timeout):
                 # no more bytes available
                 break
-            except BlockingIOError:
-                # no more bytes available
+            except OSError:
+                # UDP protocol was used and > 4096 bytes in socket buffer. This discards also the rest of the packet.
                 break
+            if self._assert_addr and addr != self._address:
+                _logger.warning(f"Received data from address %s while expected data only from %s!", addr, self._address)
+                del b
+                continue
             if not b:
                 # end of stream
                 break
 
 
+class QMI_UdpTransport(QMI_UdpTcpTransportBase):
+    """
+    An instance of QMI_UdpTransport represents a server-side UDP connection with
+    a listening port to an instrument. Client-side UDP connections are not supported.
+
+    The maximum UDP packet size is 64KB (including headers). For `write` and `read` functions,
+    this means limitations. The write function we can check the data size to send and split into
+    multiple messages. For `read` functions this is not possible and the client must limit the
+    data size as necessary. If for `read` functions the receivable data size is larger than
+    the size defined in `recvfrom(<size>)`, an exception is raised.
+    """
+
+    def __init__(self, host: str, port: int) -> None:
+        """Initialize the TCP transport by connecting to the specified address.
+
+        Parameters:
+            host: The server IP address.
+            port: The port for the address.
+        """
+        self._validate_udp_port(port)
+        super().__init__(host, port, True)
+
+    @staticmethod
+    def _validate_udp_port(port):
+        if port == QMI_Context.DEFAULT_UDP_RESPONDER_PORT:
+            raise QMI_TransportDescriptorException(f"UDP port number {port} not allowed")
+
+    def __str__(self) -> str:
+        remote_addr = format_address_and_port(self._address)
```

Comment:
this magic method doesn't have a test, but I don't know if that's important to test